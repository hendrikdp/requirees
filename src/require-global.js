const root = typeof self !== 'undefined' ? self : global;
const constants = {
    reProtocolAndHost: /^(https?:)?\/\/.+?\//i,
    reComments: /\/\*[\s\S]*?\*\/|([^:"'=]|^)\/\/.*$/mg,
    reCjsRequireCalls: /[^.]\s*require\s*\(\s*["']([^'"\s]+)["']\s*\)/g,
    reExtension: /\.(\w{2,4})$/i,
    reVersionNumber: /^\s*([*<>^~])?(\d+)\.(\d+)(\.\d+)?(\.\d+)?(\-[\w\d]*)?(\-default)?\s*$/,
    reVersionNumberInUrl: /\/(\d*\.\d+(.\d*)?(.\d*)?(\-[\w\d]*)?)\//,
    reToleranceCharacters: /^[\^~*]/,
    reVersionSplitters: /[.-]/,
    reUrlWithoutProtocolNorSpecialCharacters: /(https?:)|[\._\/:\?&=]/g,
    registryElementAttributeKeys: ['version', 'url', 'urls', 'name', 'sort', 'versions', 'type', 'factory', 'dependencies'],
    versionFormat: ['major', 'minor', 'patch', 'build', 'rc'],
    returnDefaultOnVersionStr: ['default','anonymous'],
    toleranceFormat: ['*','^','~'],
    events: {
        ns: 'requirees.',
        pre: 'pre-',
        register: 'register',
        define: 'define',
        loadFile: 'file-load',
        wireTapEventName: 'requirees.wiretaps',
        resolve: {
            regexp: /^([\w\.]*)(\[([0-9]*)\])?$/,
            fnName: 1,
            fnIndex: 3
        }
    }
};

export {root, constants};
