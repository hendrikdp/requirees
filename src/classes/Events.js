import {constants} from '../require-global.js';

export default class{

    constructor(){
        this.register = {};
    }

    subscribe(evt, callback){
        if(!(this.register[evt] instanceof Array)) this.register[evt] = [];
        this.register[evt].push(callback);
        return `${evt}[${this.register[evt].length - 1}]`;
    }

    unsubscribe(key){
        try {
            const keyFragments = key.match(constants.events.resolve.regexp);
            const fnName = keyFragments[constants.events.resolve.fnName];
            const fnIndex = keyFragments[constants.events.resolve.fnIndex];
            if(typeof fnIndex === 'undefined'){
                delete this.register[fnName];
            }else{
                delete this.register[fnName][fnIndex];
            }
        }catch(e){
            console.warn(`RequireEs: we could not unsubscribe from ${key}`, e);
        }
    }

    publish(evt, payload){
        const evts = this.register[evt] instanceof Array ? this.register[evt] : [];
        const evtsAndWireTaps = evts.concat(this.register[constants.events.wireTapEventName]);
        evtsAndWireTaps.forEach(fn => {
            try{
                if(typeof fn === 'function') fn(payload, evt);
            }catch(e){
                console.error(`RequireEs: Error while executing a function in ${evt}`, e)
            }
        });
    }

    addWireTap(fn){
        return typeof fn === 'function' ? this.subscribe(constants.events.wireTapEventName, fn) : null;
    }

}