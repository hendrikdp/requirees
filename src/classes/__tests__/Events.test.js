import Events from '../Events.js';
import {constants} from "../../require-global";

describe('Test events calls', ()=>{

    let events;
    const busFn1 = data => data.busFn1 = true;
    const busFn2 = data => data.busFn2 = true;
    beforeEach(() => events = new Events());

    test('Check if events get registered - using on', ()=>{
         events.subscribe('myevent', busFn1);
         events.subscribe('myevent', busFn2);
         expect(events.register.myevent.length).toBe(2);
    });

    test('Check if events can get executed with 1 function', ()=>{
        events.subscribe('myevent', busFn1);
        const evtPayload = {};
        events.publish('myevent', evtPayload);
        expect(evtPayload.busFn1).toBeDefined();
    });

    test('Check if events with multiple functions', ()=>{
        events.subscribe('myevent', busFn1);
        events.subscribe('myevent', busFn2);
        const evtPayload = {};
        events.publish('myevent', evtPayload);
        expect(evtPayload.busFn1).toBeDefined();
        expect(evtPayload.busFn2).toBeDefined();
    });

    test('Check if event-functions can get unsubscribed', ()=>{
        const evt1 = events.subscribe('myevent', busFn1);
        events.subscribe('myevent', busFn2);
        events.unsubscribe(evt1);
        const evtPayload = {};
        events.publish('myevent', evtPayload);
        expect(evtPayload.busFn1).toBeUndefined();
        expect(evtPayload.busFn2).toBeDefined();
    });

    test('Check if events can be completely removed', ()=>{
        events.subscribe('myevent', busFn1);
        events.subscribe('myevent', busFn2);
        events.unsubscribe('myevent');
        const evtPayload = {};
        events.publish('myevent', evtPayload);
        expect(evtPayload.busFn1).toBeUndefined();
        expect(evtPayload.busFn2).toBeUndefined();
    });

    test('Check if invalid unsubsribes do silently fail', ()=>{
        global.console = {warn: jest.fn()};
        events.unsubscribe({thisShouldNotWork: true});
        expect(console.warn).toBeCalled();
    });

    test('Check if non-functions get ignored', ()=>{
        events.subscribe('myevent', {thisShouldNotExecute: true});
        events.subscribe('myevent', busFn1);
        const evtPayload = {};
        events.publish('myevent', evtPayload);
        expect(evtPayload.busFn1).toBeDefined();
    });

    test('Check if errors in the bus, do not stop next functions to run', ()=>{
        global.console = {error: jest.fn()};
        events.subscribe('myevent', data => {throw 'If you are happy and you know it, system error!!'});
        events.subscribe('myevent', busFn1);
        const evtPayload = {};
        events.publish('myevent', evtPayload);
        expect(console.error).toBeCalled();
        expect(evtPayload.busFn1).toBeDefined();
    });

    test('Add wiretap to spy on all events', ()=>{
        const wireTap = jest.fn();
        events.addWireTap(wireTap);
        events.publish('myevent', {foo:'bar'});
        events.publish('myevent2', {foo2:'bar2'});
        expect(wireTap.mock.calls).toMatchSnapshot();
    })

    test('Check if only function can be added to the wiretaps', ()=>{
        events.addWireTap({foo:'bar'});
        expect(events.register[constants.events.wireTapEventName]).toBeUndefined();
        events.addWireTap(()=>{});
        expect(events.register[constants.events.wireTapEventName]).toBeDefined();
    })

});