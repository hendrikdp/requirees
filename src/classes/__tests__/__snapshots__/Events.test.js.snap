// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test events calls Add wiretap to spy on all events 1`] = `
[
  [
    {
      "foo": "bar",
    },
    "myevent",
  ],
  [
    {
      "foo2": "bar2",
    },
    "myevent2",
  ],
]
`;
